File: Sources/TranslateCore/UnofficialGoogleProvider.swift
import Foundation

/// 零配置 Default Translation Provider，对接 Google Translate 消费端接口。
///
/// 可用性不受保证；失败必须作为 Translation Failure 展示，不得切换提供方。
public struct UnofficialGoogleProvider: TranslationProvider {
    public let id = "unofficial-google"

    private let http: any HTTPPerforming
    private let endpoint: URL

    public init(
        http: any HTTPPerforming = URLSessionHTTPClient(),
        endpoint: URL = URL(string: "https://translate.googleapis.com/translate_a/single")!
    ) {
        self.http = http
        self.endpoint = endpoint
    }

    public func translate(_ request: TranslationRequest) -> AsyncStream<TranslationEvent> {
        AsyncStream { continuation in
            let task = Task {
                do {
                    let data = try await send(request)
                    let text = try Self.parseTranslatedText(from: data)
                    continuation.yield(.update(text))
                    continuation.yield(.completed)
                } catch {
                    continuation.yield(.failure(Self.failureMessage(for: error)))
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// 把 Google `translate_a/single` 的 JSON 数组拼成一条译文。
    /// - Parameter data: 响应体。允许前缀 `)]}'`。
    /// - Returns: 各句译文按原文顺序拼接的结果。
    public static func parseTranslatedText(from data: Data) throws -> String {
        var payload = data
        if let raw = String(data: data, encoding: .utf8) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.hasPrefix(")]}'") {
                let index = trimmed.index(trimmed.startIndex, offsetBy: 4)
                payload = Data(trimmed[index...].utf8)
            }
        }

        guard let root = try JSONSerialization.jsonObject(with: payload) as? [Any],
              let sentences = root.first as? [Any]
        else {
            throw UnofficialGoogleError.unreadableResponse
        }

        var translated = ""
        for sentence in sentences {
            guard let row = sentence as? [Any] else { continue }
            if let piece = row.first as? String {
                translated += piece
            }
        }

        let result = translated.trimmingCharacters(in: .whitespacesAndNewlines)
        if result.isEmpty {
            throw UnofficialGoogleError.emptyTranslation
        }
        return translated
    }

    private func send(_ request: TranslationRequest) async throws -> Data {
        var urlRequest = URLRequest(url: endpoint)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue(
            "application/x-www-form-urlencoded;charset=UTF-8",
            forHTTPHeaderField: "Content-Type"
        )
        urlRequest.setValue("Mozilla/5.0", forHTTPHeaderField: "User-Agent")

        var components = URLComponents()
        components.queryItems = [
            URLQueryItem(name: "client", value: "gtx"),
            URLQueryItem(name: "sl", value: request.detectedLanguage ?? "auto"),
            URLQueryItem(name: "tl", value: request.targetLanguage),
            URLQueryItem(name: "dt", value: "t"),
            URLQueryItem(name: "q", value: request.sourceText),
        ]
        urlRequest.httpBody = components.percentEncodedQuery?.data(using: .utf8)

        let (data, response) = try await http.data(for: urlRequest)
        if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
            throw UnofficialGoogleError.httpStatus(httpResponse.statusCode)
        }
        return data
    }

    private static func failureMessage(for error: Error) -> String {
        if let error = error as? UnofficialGoogleError {
            return error.localizedDescription
        }
        return "Unofficial Google Provider 请求失败。"
    }
}

enum UnofficialGoogleError: LocalizedError {
    case httpStatus(Int)
    case emptyTranslation
    case unreadableResponse

    var errorDescription: String? {
        switch self {
        case .httpStatus(let code):
            return "Unofficial Google Provider 请求失败（HTTP \(code)）。"
        case .emptyTranslation:
            return "Unofficial Google Provider 返回了空译文。"
        case .unreadableResponse:
            return "Unofficial Google Provider 返回了无法解析的响应。"
        }
    }
}

/// 未配置时的 Default Translation Provider。
public enum DefaultTranslationProvider {
    public static func unconfigured() -> any TranslationProvider {
        UnofficialGoogleProvider()
    }
}

File: lib/src/translation/providers/unofficial_google_provider.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import '../translation_types.dart';

/// 零配置翻译源；每个流独占 HTTP 连接，取消订阅即中止网络请求。
class UnofficialGoogleProvider implements TranslationProvider {
  /// endpoint 为翻译接口 URL；构造可取消的 Google Provider。
  UnofficialGoogleProvider({
    this.endpoint = 'https://translate.googleapis.com/translate_a/single',
  });

  @override
  final String id = 'unofficial-google';
  final String endpoint;

  /// request 包含原文和语言方向；返回可取消的译文、完成或失败事件流。
  @override
  Stream<TranslationEvent> translate(TranslationRequest request) {
    late final HttpClient client;
    late final StreamController<TranslationEvent> events;
    events = StreamController<TranslationEvent>(
      onListen: () async {
        client = HttpClient();
        try {
          final connection = await client.postUrl(Uri.parse(endpoint));
          connection.headers.set(
            'Content-Type',
            'application/x-www-form-urlencoded;charset=UTF-8',
          );
          connection.headers.set('User-Agent', 'Mozilla/5.0');
          connection.write(
            Uri(
              queryParameters: {
                'client': 'gtx',
                'sl': request.detectedLanguage ?? 'auto',
                'tl': request.targetLanguage,
                'dt': 't',
                'q': request.sourceText,
              },
            ).query,
          );
          final response = await connection.close();
          final body = await utf8.decodeStream(response);
          if (events.isClosed) return;
          if (response.statusCode < 200 || response.statusCode >= 300) {
            events.add(
              TranslationFailure(
                'Unofficial Google Provider 请求失败（HTTP ${response.statusCode}）。',
              ),
            );
            return;
          }
          // 消费端共享同一译文事件契约，HTTP 解析不进入会话或界面层。
          events.add(TranslationUpdate(parseTranslatedText(body)));
          events.add(const TranslationCompleted());
        } on FormatException {
          if (!events.isClosed) {
            events.add(
              const TranslationFailure(
                'Unofficial Google Provider 返回了无法解析的响应。',
              ),
            );
          }
        } catch (_) {
          // 取消导致的断连不向已经结束的会话发送失败消息。
          if (!events.isClosed) {
            events.add(
              const TranslationFailure('Unofficial Google Provider 请求失败。'),
            );
          }
        } finally {
          client.close(force: true);
          if (!events.isClosed) await events.close();
        }
      },
      onCancel: () {
        // Stream 的取消直接关闭当前请求，无需另建取消令牌或共享客户端。
        client.close(force: true);
        if (!events.isClosed) unawaited(events.close());
      },
    );
    return events.stream;
  }

  /// raw 为 Google JSON 响应；返回拼接译文，空白或结构不合法时抛出 FormatException。
  static String parseTranslatedText(String raw) {
    var payload = raw.trim();
    if (payload.startsWith(")]}'")) {
      payload = payload.substring(4);
    }
    final decoded = jsonDecode(payload);
    if (decoded is! List || decoded.isEmpty || decoded.first is! List) {
      throw const FormatException('unreadable');
    }
    final sentences = decoded.first as List;
    final buffer = StringBuffer();
    for (final sentence in sentences) {
      if (sentence is List && sentence.isNotEmpty && sentence.first is String) {
        buffer.write(sentence.first);
      }
    }
    final text = buffer.toString();
    if (text.trim().isEmpty) {
      throw const FormatException('empty');
    }
    return text;
  }
}
